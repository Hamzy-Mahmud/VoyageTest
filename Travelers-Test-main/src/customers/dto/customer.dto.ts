import { IsString, Min, Max, IsInt } from 'class-validator';

export class CustomerDto {
  @IsString()
  name: string;

  @IsInt()
  @Min(1000000000)
  @Max(9999999999)
  telephone: number;
}
