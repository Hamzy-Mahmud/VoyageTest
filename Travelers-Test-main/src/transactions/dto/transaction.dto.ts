import { IsInt, IsNumber} from 'class-validator';

export class TransactionDto {
  @IsInt()
  customer_id: number;

  @IsNumber()
  amount: number;
}