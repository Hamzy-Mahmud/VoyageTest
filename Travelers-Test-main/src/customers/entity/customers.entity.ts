import { Transaction } from 'src/transactions/entity/transactions.entity';
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  CreateDateColumn,
  OneToMany,
} from 'typeorm';

@Entity()
export class Customer {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  telephone: number;

  @CreateDateColumn()
  created_date: Date;

  @OneToMany(() => Transaction, (transaction) => transaction.customer_id)
  transactions?: Transaction[];
}
