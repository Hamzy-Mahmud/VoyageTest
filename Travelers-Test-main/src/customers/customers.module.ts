import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { CustomersController } from './controller/customers.controller';
import { Customer } from './entity/customers.entity';
import { CustomersService } from './service/customers.service';

@Module({
  imports: [TypeOrmModule.forFeature([Customer])],
  providers: [CustomersService],
  controllers: [CustomersController],
  exports: [CustomersService],
})
export class CustomersModule {}
