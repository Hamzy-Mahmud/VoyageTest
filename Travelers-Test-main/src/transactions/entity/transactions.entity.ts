import { Customer } from 'src/customers/entity/customers.entity';
import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, ManyToOne } from 'typeorm';

@Entity()
export class Transaction {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => Customer, (customer) => customer.id)
  customer_id: number;

  @Column()
  amount: number;

  @CreateDateColumn()
  created_date: Date;

}