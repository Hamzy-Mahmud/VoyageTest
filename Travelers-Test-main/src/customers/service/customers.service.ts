import { Injectable, HttpStatus, HttpException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Transaction } from 'src/transactions/entity/transactions.entity';
import { Repository, DataSource } from 'typeorm';
import { CustomerDto } from '../dto/customer.dto';
import { Customer } from '../entity/customers.entity';

@Injectable()
export class CustomersService {
  constructor(
    @InjectRepository(Customer)
    public  customerRepository: Repository<Customer>,
    private dataSource: DataSource,
  ) {}

  async create(customer: CustomerDto): Promise<Customer> {
    return await this.customerRepository.save(customer);
  }

  async findAll(): Promise<Customer[]> {
    return await this.dataSource
      .getRepository(Customer)
      .createQueryBuilder('customer')
      .orderBy('id', 'DESC')
      .take(10)
      .getMany();
  }

  async findOne(id: number): Promise<Customer> {
    return await this.customerRepository.findOneBy({ id });
  }

  async findCustomerTransactions(id: number): Promise<Customer> {
    try {
      const customerInformation = await this.dataSource
      .createQueryBuilder(Customer, 'customer')
      .leftJoinAndSelect(
        'customer.transactions',
        'transactions',
        'transactions.customer_id = :id',
        { id },
      )
      .where('transactions.customer_id = :id', { id })
      .orderBy('transactions.id', 'DESC')
      .getOne();

    return this.mapFindCustomerTransactionsResponse(customerInformation);
      
    } catch (error) {
      throw new HttpException({
        status: HttpStatus.BAD_REQUEST,
        error: 'customer_id should exists',
      }, HttpStatus.BAD_REQUEST);
    }
  }

  private mapFindCustomerTransactionsResponse(
    customerInformation: Customer,
  ): Customer {
    customerInformation.transactions = customerInformation.transactions.slice(
      0,
      10,
    );

    return customerInformation;
  }
}
