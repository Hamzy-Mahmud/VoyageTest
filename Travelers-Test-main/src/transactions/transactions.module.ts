import { Module } from '@nestjs/common';
import { TransactionsService } from './service/transactions.service';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Transaction } from './entity/transactions.entity';
import { TransactionsController } from './controller/transactions.controller';
import { CustomersModule } from 'src/customers/customers.module';

@Module({
  imports: [ CustomersModule, TypeOrmModule.forFeature([Transaction])],
  providers: [TransactionsService],
  controllers: [TransactionsController],
})
export class TransactionsModule {}
