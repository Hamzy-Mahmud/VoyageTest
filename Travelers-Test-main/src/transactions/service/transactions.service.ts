import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { CustomersService } from 'src/customers/service/customers.service';
import { Repository, DataSource } from 'typeorm';
import { TransactionDto } from '../dto/transaction.dto';
import { Transaction } from '../entity/transactions.entity';

@Injectable()
export class TransactionsService {
  constructor(
    @InjectRepository(Transaction)
    private transactionRepository: Repository<Transaction>,
    private customerService:CustomersService,
  ) {}

  async create(transaction: TransactionDto): Promise<Transaction> {
    const customer = await this.customerService.findOne(transaction.customer_id);
    if (customer) {
      return await this.transactionRepository.save(transaction);
    }
    throw new HttpException({
      status: HttpStatus.BAD_REQUEST,
      error: 'customer_id should exists',
    }, HttpStatus.BAD_REQUEST);
  }
}
