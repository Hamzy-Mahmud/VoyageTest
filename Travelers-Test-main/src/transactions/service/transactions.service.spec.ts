import { Test, TestingModule } from '@nestjs/testing';
import { TransactionsService } from './transactions.service';
import { CustomersModule } from 'src/customers/customers.module';
import { TypeOrmModule, getRepositoryToken } from '@nestjs/typeorm';
import { Transaction } from '../entity/transactions.entity';
import { CustomersService } from 'src/customers/service/customers.service';
import { TransactionDto } from '../dto/transaction.dto';
import { Repository } from 'typeorm';
import { HttpException } from '@nestjs/common';

export type MockType<T> = {
  [P in keyof T]?: jest.Mock<{}>;
};

export const repositoryMockFactory: () => MockType<Repository<any>> = jest.fn(
  () => ({
    findOne: jest.fn((entity) => entity),
    save: jest.fn().mockResolvedValue({
      customer_id: 1,
      amount: -3736699898,
      id: 1,
      created_date: new Date('2022-09-09T17:43:26.000Z'),
    }),
  }),
);

describe('TransactionsService', () => {
  let service: TransactionsService;
  let repositoryMock: MockType<Repository<Transaction>>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        CustomersModule,
        TypeOrmModule.forRoot({
          type: 'sqlite',
          database: 'travelers',
          entities: [__dirname + '/**/*.entity{.ts,.js}'],
          synchronize: true,
        }),
        TypeOrmModule.forFeature([Transaction]),
      ],
      providers: [
        TransactionsService,
        {
          provide: CustomersService,
          useValue: {
            findOne: jest.fn().mockResolvedValue({
              name: 'test 1',
              telephone: 4958173447,
              id: 1,
              created_date: new Date('2022-09-11T07:20:19.328Z'),
            }),
          },
        },
        {
          provide: getRepositoryToken(Transaction),
          useFactory: repositoryMockFactory,
        },
      ],
    }).compile();

    service = module.get<TransactionsService>(TransactionsService);
    repositoryMock = module.get(getRepositoryToken(Transaction));
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should create a transaction ', async () => {
    const body: TransactionDto = {
      customer_id: 1,
      amount: -3736699898,
    };

    const createresponse = await service.create(body);
    expect(createresponse).toMatchSnapshot();
  });
});
