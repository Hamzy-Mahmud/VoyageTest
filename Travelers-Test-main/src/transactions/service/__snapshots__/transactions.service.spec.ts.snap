// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsService should create a transaction  1`] = `
Object {
  "amount": -3736699898,
  "created_date": 2022-09-09T17:43:26.000Z,
  "customer_id": 1,
  "id": 1,
}
`;
